Speical treatment to build AOTInductor with cuda-12 from Meta internal

hl475 opened the pull request:
Differential Revision: D49042577




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

hl475 commented:
@pytorchbot label "topic: not user facing"

bertmaher commented on torch/_inductor/codecache.py:
```diff
@@ -877,9 +877,16 @@ def get_include_and_linking_paths(
+    static_link_libs = ""
+    if aot_mode and config.is_fbcode():
+        # For Meta internal cuda-12, it is recommended to static link cudarrt
+        static_link_libs = " ".join(
```
might be marginally cleaner to keep this as a list here (instead of joining) and then join with libs below like:
```
libs = " ".join(static_link_libs + ["-l" + p for p in libs])
```

bertmaher replied:
Really nbd either way though

bertmaher commented on torch/utils/cpp_extension.py:
```diff
@@ -1206,6 +1206,15 @@ def library_paths(cuda: bool = False) -> List[str]:
                 # Note that it's also possible both don't exist (see
                 # _find_cuda_home) - in that case we stay with 'lib64'.
                 lib_dir = 'lib'
+                # This is a special treatment for Meta internal cuda-12 where all libs
+                # are in lib/cuda-12 and lib/cuda-12/stubs
+                for root, dirs, files in os.walk(_join_cuda_home(lib_dir)):
+                    if 'libcudart_static.a' in files:
+                        lib_dir = os.path.join(lib_dir, root)
+                        break
+                if lib_dir != 'lib':
+                    paths.append(_join_cuda_home(lib_dir))
+                    lib_dir = os.path.join(lib_dir, "stubs")
```
Actually, this part makes me kind of nervous since it adds some meta-specific assumptions to an otherwise generic path.  Is there a way to make aot-inductor work without this hack?

hl475 replied:
Thanks Bert! I moved the change in `cpp_extension.py` back to `codecache.py` so we can use the gating `config.is_fbcode()`

aakhundov commented on torch/_inductor/codecache.py:
```diff
@@ -882,9 +893,14 @@ def get_include_and_linking_paths(
 
+    static_link_libs = []
+    if aot_mode and config.is_fbcode():
+        # For Meta internal cuda-12, it is recommended to static link cudarrt
```
Nit: "cudart", with 1 r.

hl475 committed:
Special treatment to build AOTInductor with cuda-12 from Meta internal (#108831)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/108831

Reviewed By: frank-wei

Differential Revision: D49042577

bertmaher reviewed (approved):
Thanks!

hl475 commented:
@pytorchbot merge
